ui/src/tool-panel/switch/geometry-switch-linking-infobox.tsx

```diff
@@ -154,7 +159,7 @@ const GeometrySwitchLinkingInfobox: React.FC<GeometrySwitchLinkingInfoboxProps>
                     return {
                         jointNumber: joint.number,
                         location: joint.location,
-                        segments: joint.matches,
+                        segments: joint.matches satisfies SwitchLinkingSegment[],
```

Comment:
Oli melko hämärää miten näitä tietorakenteita puljattiin frontissa. Bäkkäriltä saatava `SuggestedSwitch` käyttää eri tyyppiä (`SuggestedSwitchJointMatch`) olevaa kenttää, kuin sinne lähetettävä `SwitchLinkingParameters` rakenne, joka käyttää `SwitchLinkingSegment`-tyyppiä _samalle datalle_.

Tämähän on siis tietysti toiminut TS:n tyyppirajauksen vuoksi (lukuunottamatta kentännimitysvirhettä, ks yltä), mutta datanpuljausketju oli kyllä melko vaikeatulkintainen ainakin minulle. Tuon satisfiesin voi toki ottaa veks, mutta sanoisin, että tämä lisäys kyllä helpottaa seuraavaa lukijaa.

Oma tulkinta yleisesti, että on aika outoa puljata dataa ja käyttää eri tyyppiä näissä rakenteissa: Jos frontti saa "suggestedSwitch":n, niin miksei sitä vaan lähetetä suoraan bäkkärille kirjoitusta varten? Tällöin frontti toimisi vaan "hyväksyjänä" eikä säätäisi dataa sen enempää (lienee mahdollinen remppa ainakin).

Reply:
Satisfies on ihan hyvä kyllä :+1: . Olen tainnut itsekin tuota joskus ihmetellä kun eri tyyppejä sotketaan keskenään. Toisaalta ymmärrettävää että tuntuu tyhmältä tehdä joku mappausfunktiokaan joka vain kopioi kentät samoille nimille. Taitaa olla vähän kumsi-kumsa tämä. Kommentoin vähän tämän pohjalta, mutta käymättä tässä nyt läpi tuota backarin APIa kunnolla.

Yleisesti noissa apeissa on pyritty pitämään tallennusoliot erillisinä koska niillä on taipumus elää hiukan eri tarpeilla, mikä usein johtaa olioihin joissa on paljon nullable kenttiä joista käytetään eri osajoukkoja eri tilanteissa. Se että tallennus-API käyttää eri oliota kuin pyyntö-API pitää huolen että molemmat palvelee juuri sitä tarvetta. 

Toinen ohjaava periaate on ollut että API:n pitäisi sisältää vain se mitä se ehdottomasti tarvitsee, eli mahdolliset laskennat yms kannattaisi tehdä uudelleen jos hyväksyttävän vaihde-ehdotuksen voi kommunikoida jollain pienemmällä datalla. Tällöin backari ei usko fronttia jos pyyntöön mälvätään jotain ristiriitaista tai jos ehdotuksen haun ja tallentamisen välissä joku muu muutos vaikuttaa tilanteeseen.

Toisaalta, vaihde-ehdotus on siinä määrin selkeä oma abstraktionsa että sitä vois olla perusteltua käyttää sellaisenaan (tai niin että tallennuspyyntö sisältää sen). Historiallinen peruste erottelulle on ehkä se että aiemmin on puhuttu myös vaihdepisteiden manuaalisesta sijoittamisesta erikoistapauksiin. Tuo taitaa olla tarpeettomana kuollut ajatus jo tässä kohtaa.